### ISSUE TITLE
[environments] Support PEX executables

### ISSUE THREAD

**Issue by edoakes** [ray-project/ray#15518]
This is a feature request for supporting PEX in the runtime_env of ray, which is similar to how conda is supported (see issue #14524 2).

For example, we can package an actor “foo” in a PEX file “my_package.pex”:

@ray.remote class Foo(object): ...

And then initialize ray with the PEX environment to enable the worker to run “foo” in the PEX environment:
ray.init(runtime_env = { "pex": "path/to/my_package.pex" });

We have an example repo GitHub - leozc/fp20 1 to package ray actors into PEX files and run the actor in a local ray cluster. The commands are:

build ray_demo.pex in dist:
./pants package ray_demo:ray_demo

run ray workers in PEX environment:
./dist/ray_demo/ray_demo.pex

What we would like to have is to execute the code in a real cluster (e.g. cloud, k8s, etc.).

**Comment by ericl** [comment]
Long term it seems likely we're going to have a long tail of extra features for runtime_env. One way of supporting this to make it easy to make runtime_env pluggable. This would involve at least:

1. The ability to easily add new resource types in runtime_env (e.g., updating a dict in runtime_env.py similar to how we have a dict of node providers for autoscaler).
2. A storage API that plugins can use to save data (e.g., the PEX file) for distribution to the cluster.

cc @iycheng

**Comment by iycheng** [comment]
> Long term it seems likely we're going to have a long tail of extra features for runtime_env. One way of supporting this to make it easy to add new keys to runtime_env. This would involve at least:
> 
> 1. The ability to easily add new resource types in runtime_env (e.g., updating a dict in runtime_env.py similar to how we have a dict of node providers for autoscaler).
> 2. A storage API that plugins can use to save data (e.g., the PEX file) for distribution to the cluster.
> 
> cc @iycheng

Yes, this is what I'm currently working on. Hopefully, I'll get the PR out soon.

**Comment by mingshi-wang** [comment]
@iycheng Hi Yi, if you could share some pointers on the approach we would be happy to work on this and contribute back!

**Comment by mingshi-wang** [comment]
@iycheng Hi Yi, I wonder if you are open for a discussion on this feature request?

**Comment by iycheng** [comment]
Hi @mingshi-wang  sorry, I missed your message. Yes, I'm ok with some discussion of this one. I think right now, @architkulkarni  also has some thoughts about this one, please involve him as well.

**Comment by yehoshoua** [comment]
PEX support should be very usefully. i see this request open 6 months ago did you have any estimations ?

**Comment by yijianli-autra** [comment]
any updates on this? or can we contribute to this feature?

**Comment by architkulkarni** [comment]
We don't have a timeline for adding this, but it should be possible to write your own "third-party runtime env plugin" for this.  We don't have documentation for the plugin API at the moment and it is still experimental, but the following test https://github.com/ray-project/ray/blob/2bedf2490d7b11c2728c3fc8c57cf337a5e296ef/python/ray/tests/test_runtime_env_plugin.py has some example plugins.

If the plugin has enough interest we can consider merging it into the Ray repo.  cc @scv119

**Comment by smit-kiri** [comment]
Any updates on this? We're working with pants / PEX and would love to see this being supported!

**Comment by dvreed77** [comment]
I was able to make some progress on this and intend to contribute back in the next month or two

**Comment by marco-tortolani-datadog** [comment]
A workaround I used was adding the Pex to a docker image, create a virtual environment from the Pex, and then add the virtual environment it to the `PATH`. This way, your source code and packages are available on the Ray workers even if you only run the Pex entrypoint on the Ray head. Similar to these [Pex docs](https://pex.readthedocs.io/en/v2.1.63/recipes.html#pex-app-in-a-container), the Dockerfile looks like:
```
  "FROM python:3.10-slim",
  "COPY ray_job.pex /ray_job.pex",
  "RUN PEX_TOOLS=1 /ray_job.pex venv /venv --collisions-ok",
  'ENV PATH="/venv/bin:$PATH"',
```
If you're using [Pantsbuild](https://www.pantsbuild.org/), the setup is pretty similar to [this guide](https://www.pantsbuild.org/blog/2022/08/02/optimizing-python-docker-deploys-using-pants). This is definitely less intuitive than just setting the Pex as a runtime env directly, I'd love to hear the current status for Pex runtime plugin support!
